pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTElement.java

```diff
@@ -73,13 +244,66 @@ public void validate(@Nonnull ModelValidator validator) {
         // No-op
     }
 
+    protected static void validate(@Nonnull ModelValidator validator, @CheckForNull ModelASTMarkerInterface item) {
+        if (item != null) {
+            item.validate(validator);
+        }
+    }
+
+    protected static <T extends ModelASTMarkerInterface> void validate(@Nonnull ModelValidator validator, @CheckForNull List<T> list) {
+        if (list != null) {
+            list.forEach( item -> item.validate(validator));
+        }
+    }
+
+    protected static <K extends  ModelASTMarkerInterface, V extends  ModelASTMarkerInterface> void validate(@Nonnull ModelValidator validator, @CheckForNull Map<K, V> map) {
+        if (map != null) {
+            for (Map.Entry<K, V> entry : map.entrySet()) {
+                entry.getKey().validate(validator);
+                entry.getValue().validate(validator);
+            }
+        }
+    }
+
     /**
      * Removes the source location value from this element.
      */
     public void removeSourceLocation() {
         sourceLocation = null;
     }
 
+    /**
+     * Removes the source location value from this element.
+     */
+    protected static void removeSourceLocation(@CheckForNull ModelASTMarkerInterface item) {
```

Comment:
Could you instead change the original `removeSourceLocation` to something like:

```
protected static void removeSourceLocation(ModelASTMarkerInterface... items) {
    sourceLocation = null;
    if (items != null) { // If someone calls `removeSourceLocation(nullArg)`
        for (ModelASTMarkerInterface item : items) {
            if (item != null) {
                item.removeSourceLocation();
            }
        }
    }
}
```
And then delete the new method? That way code like this:
```
super.removeSourceLocation();
removeSourceLocation(agentType);
removeSourceLocation(variables);
```
Would just become this:
```
super.removeSourceLocation(agentType, variables);
```
I think existing callers of `removeSourceLocation()` would be fine, although I'm not sure.

Reply:
@dwnusbaum  
I like this in principle.
For anything the derives directly from element this would be fine.
But for deeper derived classes, we still need to call `super.removeSourceLocation()` to traverse up the class tree clearing source locations at each level right?  

I I like the `ModelASTMarkerInterface...  items` though.

Reply:
> But for deeper derived classes, we still need to call super.removeSourceLocation() to traverse up the class tree clearing source locations at each level right?

I think as long as there was only one `removeSourceLocation(...)` method, and every implementor of that method called `super.removeSourceLocation(...)` concatenating passed arguments with any additional fields from the current class, it would work, but the need to concatenate args would probably make it more complicated that what you have right now, so it seems best to just ignore my suggestion.